### DISCUSSION TITLE
JS embed to uWebSockets (C++)

### DISCUSSION THREAD

**Discussion by uasan** [uNetworking/uWebSockets.js#448] [Ideas]
Hello.

Many HTTP server implement embed script language:

1. [Nginx - JS](https://nginx.org/en/docs/njs/index.html)
2. [HapRoxy - Lua](https://www.haproxy.com/blog/5-ways-to-extend-haproxy-with-lua/)

This is a good solution for uWebSockets.js use cases.
The Node.js stack is not needed for a thin backend and microservices, it is enough to have a basic javascript to implement a business logic - send request to the database and communicate with other network services (websocket publish, Redis, etc).

[QuickJS](https://bellard.org/quickjs/) - is a small and embeddable Javascript engine, this is an excellent implementation JS from Fabrice Bellard, is a French computer programmer known for writing FFmpeg, QEMU.

What do you think?
Thanks.

**Comment by ghost** [comment]
I have thought about it many many times. But the gains are way too small in comparison with what you lose in terms of market. Nodejs is the creation of an insane group of morons, but it also has a huge market, so playing along with nodejs is still better than to try and introduce yet another JS server. A standard can only be a standard if people stop making new standards. And honestly working with this is going to kill me dead from a bursting artery in my brain. In fact uWS.js can already run as a stand-alone js server

**Reply by uasan** [reply]
> But the gains are way too small in comparison with what you lose in terms of market. Nodejs

Yes, I understand that it is much easier to sell uWS.js, but real users really do not really need Node.js, they need JS to easily and quickly write business logic.

Ok, thanks.

**Reply by ghost** [reply]
Possibly I could update the stand-alone mode and start building stand-alone binaries for people to try, but I can already guarantee people will be complaining about how it is not like Nodejs. I can already see how people will be assuming they can just download 57 TB of npm garbage and when it fails to run they'll report it as a bug and want me to make it Nodejs

**Reply by uasan** [reply]
These people will continue to use uWS.js (as Node.js module)
But if they want to move to the next level they will have a standalone uWS.

**Reply by slavaGanzin** [reply]
If you ever tried to write something in limited environments you wouldn't be so bullish on this. I have experience with Lua and embeddable JS in Redis and Nginx. So believe me: you don't want to do this
Even if it doesn't support some minor features you would be forced to write everything by yourself. Because nobody optimizes NPM modules for strict environments. And that's kills main purpose of using Node.js: packages. Sometimes shitty, sometimes bloated, but working.

And if you want to write everything by yourself. Take a week off, learn C, it's not that hard.

**Reply by ghost** [reply]
Best case scenario would be if people used Swift, Rust, C or C++. All these languages are LLVM based and have zero cost abstractions.

I think I can keep my sanity at decent levels by focusing on expanding to these markets rather than try to reinvent yet another JS thing that will end up driving me crazy in the end.

You don't really gain anything substantial from leaving Nodejs - maybe a few percentages more req/sec but you're still way off from what you can achieve with Swift or Rust

**Reply by uasan** [reply]
Ok, your arguments are reasonable )

**Reply by ghost** [reply]
@uasan I like your idea and I would like to have it made in a perfect world, but in reality it is never ever going to happen.
